core/src/main/java/org/apache/iceberg/FilteredManifest.java

```diff
@@ -193,4 +164,29 @@ private InclusiveMetricsEvaluator metricsEvaluator() {
     }
     return lazyMetricsEvaluator;
   }
+
+  private boolean requireStatsProjection() {
+    // Make sure we have all stats columns for metrics evaluator
+    return rowFilter != Expressions.alwaysTrue() &&
+        !columns.containsAll(ManifestReader.ALL_COLUMNS) &&
+        !columns.containsAll(STATS_COLUMNS);
+  }
+
+  private boolean dropStats() {
+    // Make sure we only drop all stats if we had projected all stats
+    // We do not drop stats even if we had partially added some stats columns
+    return rowFilter != Expressions.alwaysTrue() &&
+        !columns.containsAll(ManifestReader.ALL_COLUMNS) &&
+        Sets.intersection(Sets.newHashSet(columns), STATS_COLUMNS).isEmpty();
+  }
+
+  private static Collection<String> withStatsColumns(Collection<String> columns) {
+    if (columns.containsAll(ManifestReader.ALL_COLUMNS)) {
+      return columns;
+    } else {
+      List<String> projectColumns = Lists.newArrayList(columns);
+      projectColumns.addAll(STATS_COLUMNS); // order doesn't matter
```

Comment:
Maybe it makes sense to remove duplicates here? There could be a scenario where a user projects stats columns partially and this ends up adding duplicate columns. Maybe just make `projectColumns` a set if order doesn't matter?

Reply:
It doesn't matter because `Schema.select` ends up using a hash set of field IDs.

Reply:
Thks!  I'll leave it as it is then.